worker/caasapplicationprovisioner/application.go

```diff
@@ -103,82 +108,134 @@ func (a *appWorker) loop() error {
 	app := a.broker.Application(a.name, caas.DeploymentStateful)
 
 	var appLife life.Value
 	var appChanges watcher.NotifyChannel
 	var replicaChanges watcher.NotifyChannel
+	var appStateChanges watcher.NotifyChannel
 	var lastReportedStatus map[string]status.StatusInfo
 
-	for {
-		select {
-		case <-a.catacomb.Dead():
-			return a.catacomb.Err()
-		case <-a.changes:
-			appLife, err = a.facade.Life(a.name)
-			if errors.IsNotFound(err) {
-				appLife = life.Dead
-			} else if err != nil {
+	done := false
+
+	handleChange := func() error {
+		appLife, err = a.facade.Life(a.name)
+		if errors.IsNotFound(err) {
+			appLife = life.Dead
+		} else if err != nil {
+			return errors.Trace(err)
+		}
+		switch appLife {
+		case life.Alive:
+			if appStateChanges == nil {
+				appStateWatcher, err := a.facade.WatchApplication(a.name)
+				if err != nil {
+					return errors.Annotatef(err, "failed to watch for changes to application %q", a.name)
+				}
+				err = a.catacomb.Add(appStateWatcher)
+				if err != nil {
+					appStateWatcher.Kill()
+					return errors.Trace(err)
+				}
+				appStateChanges = appStateWatcher.Changes()
+			}
+			err = a.alive(app)
+			if err != nil {
 				return errors.Trace(err)
 			}
-			switch appLife {
-			case life.Alive:
-				err = a.alive(app)
+			if appChanges == nil {
+				appWatcher, err := app.Watch()
 				if err != nil {
-					return errors.Trace(err)
+					return errors.Annotatef(err, "failed to watch for changes to application %q", a.name)
 				}
-				if appChanges == nil {
-					appWatcher, err := app.Watch()
-					if err != nil {
-						return errors.Annotatef(err, "failed to watch for changes to application %q", a.name)
-					}
-					err = a.catacomb.Add(appWatcher)
-					if err != nil {
-						appWatcher.Kill()
-						return errors.Trace(err)
-					}
-					appChanges = appWatcher.Changes()
+				err = a.catacomb.Add(appWatcher)
+				if err != nil {
+					appWatcher.Kill()
+					return errors.Trace(err)
 				}
-				if replicaChanges == nil {
-					replicaWatcher, err := app.WatchReplicas()
-					if err != nil {
-						return errors.Annotatef(err, "failed to watch for changes to replicas %q", a.name)
-					}
-					err = a.catacomb.Add(replicaWatcher)
-					if err != nil {
-						replicaWatcher.Kill()
-						return errors.Trace(err)
-					}
-					replicaChanges = replicaWatcher.Changes()
+				appChanges = appWatcher.Changes()
+			}
+			if replicaChanges == nil {
+				replicaWatcher, err := app.WatchReplicas()
+				if err != nil {
+					return errors.Annotatef(err, "failed to watch for changes to replicas %q", a.name)
 				}
-			case life.Dying:
-				err = a.dying(app)
+				err = a.catacomb.Add(replicaWatcher)
 				if err != nil {
+					replicaWatcher.Kill()
 					return errors.Trace(err)
 				}
-			case life.Dead:
-				return a.dead(app)
+				replicaChanges = replicaWatcher.Changes()
+			}
+		case life.Dying:
+			err = a.dying(app)
+			if err != nil {
+				return errors.Trace(err)
+			}
```

Comment:
Inline the err for briefity? etc

Reply:
I'm generally not a fan of inline statements but I'll do it for you :)